File: sdks/python/roost/resources/members.py
"""``roost.members`` — site membership.

Drives:

  GET    /api/sites/{siteId}/members
  POST   /api/sites/{siteId}/members             { uid | email, role }
  PATCH  /api/sites/{siteId}/members/{uid}       { role }
  DELETE /api/sites/{siteId}/members/{uid}
  POST   /api/sites/{siteId}/transfer-ownership  { successorUid }

Construct via ``roost.members(site_id)`` — each instance is bound to one
site. The per-site role is reported server-side (owner / superadmin /
admin / member) and surfaced on each ``Member.role``.

PER-SITE ROLES. ``set_role`` changes someone's standing on THIS site only,
leaving their global role untouched. Before it existed the per-site role was
derived from the global one, so "make them an admin here" meant promoting them
everywhere they belonged — the leak per-site roles close.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from roost.client import RoostClient


#: Roles that can be ASSIGNED. ``owner`` is deliberately absent — ownership
#: moves only through :meth:`Members.transfer_ownership`, which is transactional
#: and re-reads the owner inside it, so an admin can neither promote themselves
#: into ownership nor demote the owner out of it.
AddRole = Literal["member", "admin"]

#: Role as REPORTED by the API. ``owner`` and ``superadmin`` are derived
#: standings, not settings.
PerSiteRole = Literal["owner", "superadmin", "admin", "member"]


@dataclass(slots=True)
class Member:
    uid: str
    email: str | None
    role: PerSiteRole
    display_name: str | None


def _parse_member(raw: dict[str, Any]) -> Member:
    role = raw.get("role")
    if role not in ("owner", "superadmin", "admin", "member"):
        role = "member"
    return Member(
        uid=str(raw.get("uid", "")),
        email=raw.get("email"),
        role=role,  # type: ignore[arg-type]
        display_name=raw.get("displayName"),
    )


class Members:
    """Site-membership management bound to one site."""

    def __init__(self, client: "RoostClient", site_id: str) -> None:
        self._client = client
        self._site_id = site_id

    @property
    def site_id(self) -> str:
        return self._site_id

    def _base(self) -> str:
        return f"/api/sites/{self._site_id}/members"

    async def list(self) -> list[Member]:
        resp = await self._client.request(self._base())
        data = resp.data if isinstance(resp.data, dict) else {}
        return [
            _parse_member(m)
            for m in (data.get("members") or [])
            if isinstance(m, dict)
        ]

    async def add(
        self,
        uid: str | None = None,
        *,
        email: str | None = None,
        role: AddRole = "member",
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Add a member by ``uid`` OR ``email`` — exactly one, never both.

        ``email`` is the affordance an admin actually has (they know a
        colleague's address, not their uid); the server trims and lowercases it
        and resolves it through Firebase Auth to the same uid path.

        ``role="admin"`` is honoured only when the target's GLOBAL role is
        already admin/superadmin; otherwise membership is extended and the
        response carries ``roleHonored: False``. Promotion is a separate,
        explicit endpoint — never a side-effect of adding someone to a site.
        """
        if (uid is None) == (email is None):
            raise ValueError("provide exactly one of uid or email")
        body: dict[str, Any] = {"role": role}
        if uid is not None:
            body["uid"] = uid
        else:
            body["email"] = email
        resp = await self._client.request(
            self._base(),
            method="POST",
            body=body,
            idempotency_key=idempotency_key,
        )
        return resp.data if isinstance(resp.data, dict) else {}

    async def set_role(
        self,
        uid: str,
        role: AddRole,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Change a member's role on THIS site; their global role is untouched.

        Refuses the site owner with 409 ``cannot_change_owner_role`` — the
        owner's role is set by ownership, so use
        :meth:`transfer_ownership` instead.
        """
        idem = idempotency_key or f"py-sdk-members-setrole-{uuid.uuid4()}"
        resp = await self._client.request(
            f"{self._base()}/{uid}",
            method="PATCH",
            body={"role": role},
            headers={"Idempotency-Key": idem},
        )
        return resp.data if isinstance(resp.data, dict) else {}

    async def remove(
        self,
        uid: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        # Preserve a resource-specific prefix rather than the core client's
        # generic py-sdk DELETE key.
        idem = idempotency_key or f"py-sdk-members-remove-{uuid.uuid4()}"
        resp = await self._client.request(
            f"{self._base()}/{uid}",
            method="DELETE",
            headers={"Idempotency-Key": idem},
        )
        return resp.data if isinstance(resp.data, dict) else {}


    async def transfer_ownership(
        self,
        successor_uid: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Hand this site to another user, atomically.

        Only the current owner or a superadmin may do this — a site admin is
        refused with 403 ``not_owner``. The successor need not already be a
        member; they are made one by the same transaction. The outgoing owner is
        demoted to ``admin`` and KEEPS their membership: handing a site over is
        not eviction.
        """
        idem = idempotency_key or f"py-sdk-transfer-owner-{uuid.uuid4()}"
        resp = await self._client.request(
            f"/api/sites/{self._site_id}/transfer-ownership",
            method="POST",
            body={"successorUid": successor_uid},
            headers={"Idempotency-Key": idem},
        )
        return resp.data if isinstance(resp.data, dict) else {}


__all__ = ["AddRole", "Member", "Members", "PerSiteRole"]

File: sdks/node/src/resources/members.ts
/**
 * `owlette.members(siteId)` — site-scoped membership management.
 *
 *   GET    /api/sites/{siteId}/members
 *   POST   /api/sites/{siteId}/members            — body `{ uid | email, role }`
 *   PATCH  /api/sites/{siteId}/members/{uid}      — body `{ role }`
 *   DELETE /api/sites/{siteId}/members/{uid}
 *   POST   /api/sites/{siteId}/transfer-ownership — body `{ successorUid }`
 *
 * The constructor binds to a `siteId`. Exposed as a factory on the root
 * `Owlette` instance so callers do `owlette.members(siteId).list()`.
 *
 * PER-SITE ROLES. `setRole` changes someone's role on THIS site only, leaving
 * their global role untouched. Before it existed the per-site role was derived
 * from the global one, so "make them an admin here" meant promoting them
 * everywhere they belonged — the leak per-site roles close.
 */
import { randomUUID } from 'crypto';
import type { OwletteClient } from '../lib/client';

// types

/**
 * Role as REPORTED by the API. `owner` and `superadmin` are derived standings,
 * not settings — see `AssignableSiteRole` for what can actually be written.
 */
export type SiteMemberRole = 'owner' | 'superadmin' | 'admin' | 'member';

/**
 * Roles that can be ASSIGNED. `owner` is deliberately absent: ownership moves
 * only through `transferOwnership`, which is transactional and re-reads the
 * owner inside it, so an admin can neither promote themselves into ownership
 * nor demote the owner out of it.
 */
export type AssignableSiteRole = 'admin' | 'member';

export interface SiteMember {
  uid: string;
  email: string | null;
  role: SiteMemberRole;
  globalRole: string;
  displayName: string | null;
}

/**
 * Identify the target by EXACTLY ONE of `uid` or `email` — supplying both, or
 * neither, is a 400. `email` is trimmed and lowercased, then resolved through
 * Firebase Auth to the same uid path.
 */
export type AddMemberOptions = { role: AssignableSiteRole; idempotencyKey?: string } & (
  | { uid: string; email?: never }
  | { email: string; uid?: never }
);

export interface AddMemberResult {
  uid: string;
  siteId: string;
  requestedRole: AssignableSiteRole;
  /**
   * Always `true`. Roles are per-site: the requested role is written into the
   * membership document, which is what grants, so the request is always honoured.
   *
   * @deprecated Retained only so the response shape stays stable. It used to
   * report whether the target's GLOBAL role would make an `admin` request stick,
   * back when per-site roles were derived at read time. It reported `false` while
   * a real site-admin row was written.
   */
  roleHonored: boolean;
  globalRole: string;
}

export interface SetMemberRoleResult {
  siteId: string;
  uid: string;
  role: AssignableSiteRole;
}

export interface RemoveMemberResult {
  siteId: string;
  uid: string;
  wasMember: boolean;
}

export interface TransferOwnershipResult {
  siteId: string;
  previousOwnerUid: string;
  newOwnerUid: string;
}

// resource

export class Members {
  constructor(
    private readonly client: OwletteClient,
    private readonly siteId: string,
  ) {}

  private get base(): string {
    return `/api/sites/${encodeURIComponent(this.siteId)}/members`;
  }

  /**
   * Members of this site, each with their derived per-site role.
   *
   * A member's OTHER site memberships are deliberately not returned — that
   * would hand a site admin the site ids of every other organisation the person
   * belongs to. (This type previously declared a required `sites` field the API
   * has never sent.)
   */
  async list(): Promise<SiteMember[]> {
    const res = await this.client.request<{ members: SiteMember[] }>(this.base);
    return res.data.members;
  }

  async add(opts: AddMemberOptions): Promise<AddMemberResult> {
    const res = await this.client.request<AddMemberResult>(this.base, {
      method: 'POST',
      body: opts.uid !== undefined
        ? { uid: opts.uid, role: opts.role }
        : { email: opts.email, role: opts.role },
      idempotencyKey: opts.idempotencyKey ?? `sdk-members-add-${randomUUID()}`,
    });
    return res.data;
  }

  /**
   * Change a member's role on THIS site. Their global role is untouched.
   *
   * Refuses the site owner with 409 `cannot_change_owner_role` — the owner's
   * role is set by ownership, so use {@link transferOwnership} instead.
   */
  async setRole(
    uid: string,
    role: AssignableSiteRole,
    opts: { idempotencyKey?: string } = {},
  ): Promise<SetMemberRoleResult> {
    const res = await this.client.request<SetMemberRoleResult>(
      `${this.base}/${encodeURIComponent(uid)}`,
      {
        method: 'PATCH',
        body: { role },
        idempotencyKey: opts.idempotencyKey ?? `sdk-members-setrole-${randomUUID()}`,
      },
    );
    return res.data;
  }

  async remove(uid: string): Promise<RemoveMemberResult> {
    const res = await this.client.request<RemoveMemberResult>(
      `${this.base}/${encodeURIComponent(uid)}`,
      { method: 'DELETE' },
    );
    return res.data;
  }

  /**
   * Hand this site to another user, atomically.
   *
   * Only the current owner or a superadmin may do this — a site admin is
   * refused with 403 `not_owner`. The successor need not already be a member;
   * they are made one by the same transaction. The outgoing owner is demoted to
   * `admin` and KEEPS their membership: handing a site over is not eviction.
   */
  async transferOwnership(
    successorUid: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<TransferOwnershipResult> {
    const res = await this.client.request<TransferOwnershipResult>(
      `/api/sites/${encodeURIComponent(this.siteId)}/transfer-ownership`,
      {
        method: 'POST',
        body: { successorUid },
        idempotencyKey: opts.idempotencyKey ?? `sdk-transfer-owner-${randomUUID()}`,
      },
    );
    return res.data;
  }
}
